src/translate/axiom_rules.py

```diff
@@ -1,29 +1,315 @@
+def compute_simplified_axioms(axioms):
+    """Remove duplicate axioms, duplicates within axioms, and dominated axioms."""
+
     if DEBUG:
-        verify_layering_condition(axioms, axiom_init, axiom_layers)
-    return axioms, list(axiom_init), axiom_layers
+        assert len(set(axiom.effect for axiom in axioms)) == 1
+
+    # Remove duplicates from axiom conditions.
+    for axiom in axioms:
+        axiom.condition = sorted(set(axiom.condition))
+
+    # Remove dominated axioms.
+    axioms_to_skip = set()
+    axioms_by_literal = defaultdict(set)
+    for axiom in axioms:
+        if axiom.effect in axiom.condition:
+            axioms_to_skip.add(id(axiom))
+        else:
+            for literal in axiom.condition:
+                axioms_by_literal[literal].add(id(axiom))
+
+    for axiom in axioms:
+        if id(axiom) in axioms_to_skip:
+            continue   # Required to keep one of multiple identical axioms.
+        if not axiom.condition:  # empty condition: dominates everything
+            return [axiom]
+        literals = iter(axiom.condition)
+        dominated_axioms = axioms_by_literal[next(literals)].copy()
+        for literal in literals:
+            dominated_axioms &= axioms_by_literal[literal]
+        for dominated_axiom in dominated_axioms:
+            if dominated_axiom != id(axiom):
+                axioms_to_skip.add(dominated_axiom)
+    return [axiom for axiom in axioms if id(axiom) not in axioms_to_skip]
+
+
+def compute_clusters(axioms, goals, operators):
+    dependencies = AxiomDependencies(axioms)
+
+    # Compute necessary literals and prune unnecessary vars from dependencies.
+    necessary_literals = compute_necessary_literals(dependencies, goals, operators)
+    dependencies.remove_unnecessary_variables(necessary_literals)
+
+    groups = get_strongly_connected_components(dependencies)
+    clusters = [AxiomCluster(group) for group in groups]
+
+    # Compute mapping from variables to their clusters and set needed_negatively.
+    variable_to_cluster = {}
+    for cluster in clusters:
+        for variable in cluster.variables:
+            variable_to_cluster[variable] = cluster
+            if variable.negate() in necessary_literals:
+                cluster.needed_negatively = True
+
+    # Assign axioms to their clusters.
+    for axiom in axioms:
+        # axiom.effect is derived but might have been pruned
+        if axiom.effect in dependencies.derived_variables:
+            variable_to_cluster[axiom.effect].axioms[axiom.effect].append(axiom)
+
+    removed = 0
+    with timers.timing("Simplifying axioms"):
+        for cluster in clusters:
+            for variable in cluster.variables:
+                old_size = len(cluster.axioms[variable])
+                cluster.axioms[variable] = compute_simplified_axioms(cluster.axioms[variable])
+                removed += old_size - len(cluster.axioms[variable])
+    print("Simplifying axioms removed {} axioms".format(removed))
```

Comment:
This required a custom parser in lab if I remember correctly. Do we have a general mechanism to make new outputs like these available in lab in the future? This also concerns the other custom parsers used for this issue.

Reply:
As hinted at in my other comment, we could change this line to `print(f"Translator axioms removed by simplifying: {removed})` and let the standard translator parser pick this up automatically.

Reply:
Thanks for the hint! However, please let's not mix and match string formatting styles. No other part of the translator uses f-strings. Actually, no other part of the translator uses `format` either. Our current style is `print("Translator axioms removed by simplifying: %d" % removed)`. If people want to move to another way of string formatting, great -- but please not by everyone using their own preferred style. (For what it's worth, we *are* inconsistent about whether to prefer `"%s"` or `"%d"`. Switching to `format` or f-strings could help address that; they allow the same distinction, but there is a more obvious default.)